crates/bevy_ecs/src/reflect.rs

```diff
@@ -27,6 +27,61 @@ pub struct ReflectComponent {
     copy: fn(&World, &mut World, Entity, Entity),
 }
 
+/// The raw function pointers needed to make up a [`ReflectComponent`].
+///
+/// This is used when creating custom implementations of [`ReflectComponent`] with
+/// [`ReflectComponent::new()`].
+///
+/// > **Note:**
+/// > Creating custom implementations of [`ReflectComponent`] is an advanced feature that most users
+/// > will not need.
+/// > Usually you will obtain a [`ReflectComponent`] for a type by retrieving it from the
+/// > [`TypeRegistry`][bevy_reflect::TypeRegistry].
+///
+/// Creating a custom [`ReflectComponent`] may be useful if you need to create new component types
+/// at runtime, for example, for scripting implementations.
+///
+/// By creating a custom [`ReflectComponent`] and inserting it into a type's
+/// [`TypeRegistration`][bevy_reflect::TypeRegistration],
+/// you can modify the way that reflected components of that type will be inserted into the Bevy
+/// world.
+pub struct ReflectComponentFns {
+    /// Function pointer implementing [`ReflectComponent::insert()`].
+    pub insert: fn(&mut World, Entity, &dyn Reflect),
+    /// Function pointer implementing [`ReflectComponent::apply()`].
+    pub apply: fn(&mut World, Entity, &dyn Reflect),
+    /// Function pointer implementing [`ReflectComponent::apply_or_insert()`].
+    pub apply_or_insert: fn(&mut World, Entity, &dyn Reflect),
+    /// Function pointer implementing [`ReflectComponent::remove()`].
+    pub remove: fn(&mut World, Entity),
+    /// Function pointer implementing [`ReflectComponent::reflect()`].
+    pub reflect: fn(&World, Entity) -> Option<&dyn Reflect>,
+    /// Function pointer implementing [`ReflectComponent::reflect_mut()`].
+    pub reflect_mut: unsafe fn(&World, Entity) -> Option<Mut<dyn Reflect>>,
+    /// Function pointer implementing [`ReflectComponent::copy()`].
+    pub copy: fn(&World, &mut World, Entity, Entity),
+}
+
+impl ReflectComponentFns {
+    /// Get the default set of [`ReflectComponentFns`] for a specific component type using its
+    /// [`FromType`] implementation.
+    ///
+    /// This is useful if you want to start with the default implementation before overriding some
+    /// of the functions to create a custom implementation.
+    pub fn new<T: Component + Reflect + FromWorld>() -> Self {
+        let c = <ReflectComponent as FromType<T>>::from_type();
+        Self {
+            insert: c.insert,
+            apply: c.apply,
+            apply_or_insert: c.apply_or_insert,
+            remove: c.remove,
+            reflect: c.reflect,
+            reflect_mut: c.reflect_mut,
+            copy: c.copy,
+        }
```

Comment:
This also makes me wonder if `ReflectComponent` should store `ReflectComponentFns` instead of duplicating all the fields. Thoughts on that?

Reply:
That sounds reasonable to me. :+1: